### PR TITLE
New post: Optimising a simple ActiveRecord query.

### PR DESCRIPTION
**Pull request by joosto**
After writing a short blog post on Workplace about an ActiveRecord query that I optimised together with @martyphee, I was asked to turn the the post into an engineering blog post.

### CODE REVIEW

File: _posts/2018-04-12-optimising-active-record.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+The problem occurred in the query in this innocuous looking method, which returns the date of the first order of a company on Deliveroo. Apart from containing a `JOIN`, which can be expensive in certain situations, the query simply takes the first order that a company has received, and takes its date of creation.
+
+```ruby
+def self.date_of_first_order(restaurant_company)
+  OrderEvent
+    .joins(:restaurant)
+    .where(restaurants: { company_id: restaurant_company.id })
+    .first
+    &.created_at
```
Comment (tommyp) [inline _posts/2018-04-12-optimising-active-record.md]: I was going to ask if this was a typo but then I took the time to Google it and found [an explanation](https://stackoverflow.com/questions/36812647/what-does-ampersand-dot-mean-in-ruby/36812667). This is so handy!

File: _posts/2018-04-12-optimising-active-record.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    .select(:id)
+    .joins(:restaurant)
+    .where(restaurants: { company_id: restaurant_company.id })
+    .minimum(:id)
+end
+```
+
+Done! The trade off is that the resulting Ruby is more verbose and non-standard, which is why an optimisation like this shouldn’t be made prematurely. However, in this case the performance increase is worth it.
+
+### PS.
```
Comment (kapowaz) [inline _posts/2018-04-12-optimising-active-record.md]: Nitpick, but this might look better as `P.S.`?

Comment (joosto) [reply _posts/2018-04-12-optimising-active-record.md]: Thanks, updated :)

**Comment by pedrocunha** [comment]
I would be curious to see a comment from the DBA folks here. I think the problem here is mainly when a given restaurant has too many orders. I am pretty sure the sort by on a primary key especially should be extremely fast?

**Pull request merged**